### PR TITLE
raftstore-v2: actively handle ResCmdChannel after peer is destroyed

### LINKED ISSUE #14675: [Dynamic Regions] latency spike caused by PessimisticLock when running tpcc stability workload with v2
### What version of TiKV are you using?

lastest master with pr https://github.com/tikv/tikv/pull/14536
### What operating system and CPU are you using?

x86
### Steps to reproduce

run tpcc 1k warehouse for 7 days

### What did you expect?
qps and latency is stable.

### What did happened?
qps and latency spikes, caused by PessimisticLock kv request latency spike.

image
image
image

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #14675

What's Changed:

```commit-message
actively handle ResCmdChannel after peer is destroyed.
```
https://github.com/tikv/tikv/issues/14675 encounters a phenomenon where all txns block at acquiring pessimistic lock and QPS can drop to 0 forever. 

This is **caused** by that **`CmdResChannel` is dropped unexpectedly** which leads to txn task in scheduler worker **unable to release latches** infinitely.

Steps:
Asume the original peers in some time are (1, 1), (2, 2), (3, 3), (4, 4 learner), and the leader is (1, 1).
1. conf change { AddNode(4, 4), AddLearner(1, 1) }
2. prewrite (say it p1) and peer 1 proposes p1 in raftstore,
assume peer 2 and peer 3 cannot receive the MsgAppend immediately, so the command will not be committed for some while.
3. transfer leader to `IncomingVoter` peer 4
4. conf change { RemoveNode(1, 1) }, now peer 1 cannot communite with others
5. peer 1 starts elections due to timeout and receive tombstone msg
6. peer 1 destroyed itself

Now, `CmdResChannel` is still held in PeerFsm of peer 1 which will be dropped due to the destroy of PeerFsm. This is unexpected in the sense of txn task in  the scheduler worker, and thus p1 has no chance to release latches.

Furthermore, p1 holds some latches which block the following txns. These txns may also hold some latches acquired successfully but never release them (as p1 will not release the latches they wait), which may block even more txns following...

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -1597,10 +1597,14 @@ impl<E: Engine, L: LockManager> TxnScheduler<E, L> {
...
             }
         }
+
         // If it's not finished while the channel is closed, it means the write
         // is undeterministic. in this case, we don't know whether the
         // request is finished or not, so we should not release latch as
         // it may break correctness.
+        // However, not release latch will cause deadlock which may ultimately block all
+        // following txns, so we panic here.
+        panic!("response channel is unexpectedly dropped, cid {}", cid);
```
Comment (tonyxuqqi) [inline src/storage/txn/scheduler.rs:1607]: loop transaction team for review this line.

Comment (SpadeA-Tang) [reply src/storage/txn/scheduler.rs:1607]: @MyonKeminta would you mind helping for revewing this?

**Review by tonyxuqqi — APPROVED** [review]

**Comment by dbsid** [comment]
with this PR, after 5 days stability run, the issue is not reproduced.

img_v2_40d2bb1b-f6d2-4315-847d-1f1ea32c29bg

File: src/storage/txn/scheduler.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/storage/txn/scheduler.rs:1607]: I've noticed the problem last week. I'm not sure if there are any corner cases that the channel is dropped *incorrectly* in raftstore, but this panic looks quite reasonable to me.

Comment (MyonKeminta) [reply src/storage/txn/scheduler.rs:1607]: By the way consider also printing `tag` here.

Comment (SpadeA-Tang) [reply src/storage/txn/scheduler.rs]: > By the way consider also printing `tag` here.

Done

**Review by overvenus — APPROVED** [review]

**Pull request merged**